algorithms/map-optimization/src/outlier-rejection-solver.cc

```diff
@@ -47,69 +48,86 @@ double computeSquaredReprojectionError(
 }
 
 void findOutlierLandmarks(
-    const vi_map::VIMap& map, const vi_map::LandmarkIdSet& landmarks_in_problem,
+    const vi_map::VIMap& map, const vi_map::LandmarkIdSet& landmark_ids_set,
     const bool use_reprojection_error,
     const double same_mission_reprojection_error_px,
     const double other_mission_reprojection_error_px,
-    vi_map::LandmarkIdSet* outlier_landmarks) {
+    vi_map::LandmarkIdList* outlier_landmarks) {
   CHECK_NOTNULL(outlier_landmarks)->clear();
 
   const double same_mission_reproj_error_px_sq =
       same_mission_reprojection_error_px * same_mission_reprojection_error_px;
   const double other_mission_reproj_error_px_sq =
       other_mission_reprojection_error_px * other_mission_reprojection_error_px;
 
-  for (const vi_map::LandmarkId& landmark_id : landmarks_in_problem) {
-    const vi_map::Landmark& landmark = map.getLandmark(landmark_id);
-
-    const vi_map::MissionId& landmark_store_mission_id =
-        map.getLandmarkStoreVertex(landmark_id).getMissionId();
-
-    landmark.forEachObservation(
-        [&](const vi_map::KeypointIdentifier& keypoint_id) {
-          const vi_map::Vertex& observer_vertex =
-              map.getVertex(keypoint_id.frame_id.vertex_id);
-
-          const int frame_idx = keypoint_id.frame_id.frame_index;
-          CHECK(observer_vertex.isVisualFrameSet(frame_idx));
-          CHECK(observer_vertex.isVisualFrameValid(frame_idx));
-
-          CHECK_LT(
-              keypoint_id.keypoint_index,
-              observer_vertex.getVisualFrame(frame_idx)
-                  .getNumKeypointMeasurements());
-
-          const Eigen::Vector3d p_C_fi =
-              map.getLandmark_p_C_fi(landmark_id, observer_vertex, frame_idx);
-
-          if (p_C_fi[2] <= 0.0) {
-            outlier_landmarks->emplace(landmark_id);
-            return;
-          }
+  vi_map::LandmarkIdList landmark_ids(
+      landmark_ids_set.begin(), landmark_ids_set.end());
+
+  std::mutex outlier_landmarks_mutex;
+  std::function<void(const std::vector<size_t>&)> process_landmark =
+      [&](const std::vector<size_t>& batch) {
+        for (size_t item : batch) {
+          const vi_map::LandmarkId& landmark_id = landmark_ids[item];
+          const vi_map::Landmark& landmark = map.getLandmark(landmark_id);
+
+          const vi_map::MissionId& landmark_store_mission_id =
+              map.getLandmarkStoreVertex(landmark_id).getMissionId();
+
+          // Iterate over all the observations
```

Comment:
To be able to break out of the loop since if a landmark is an outlier based on one observation we don't want to check the others as well and the forEachObservation will always iterate over everything.